src/v/storage/api.cc

```diff
@@ -93,27 +93,55 @@ ss::future<> api::monitor() {
 
         auto usage = co_await disk_usage();
 
+        // when low watermark is exceeded we'll begin trying to reduce storage
+        // usage in order to avoid blocking writes
+        const auto hwm_exceeded
+          = _log_storage_max_size().has_value()
+            && usage.usage.total() > uint64_t(
+                 config::shard_local_cfg().log_storage_max_size_high_watermark()
+                 * _log_storage_max_size().value());
+
+        // when the max is actually exceeded we'll begin blocking writes
         const auto max_exceeded = _log_storage_max_size().has_value()
                                   && usage.usage.total()
                                        > _log_storage_max_size().value();
 
         // broadcast if state changed
-        if (max_exceeded != _max_size_exceeded) {
-            co_await container().invoke_on_all([max_exceeded](api& api) {
-                vlog(
-                  stlog.info,
-                  "{} maximum size exceeded flag",
-                  max_exceeded ? "Setting" : "Clearing");
-                api._max_size_exceeded = max_exceeded;
-            });
+        if (
+          max_exceeded != _max_size_exceeded
+          || hwm_exceeded != _high_watermark_exceeded) {
+            co_await container().invoke_on_all(
+              [max_exceeded, hwm_exceeded](api& api) {
+                  vlog(
+                    stlog.info,
+                    "{} maximum size exceeded flag, and {} low watermark "
+                    "exceeded flag.",
+                    max_exceeded ? "Setting" : "Clearing",
+                    hwm_exceeded ? "Setting" : "Clearing");
+                  api._max_size_exceeded = max_exceeded;
+                  api._high_watermark_exceeded = hwm_exceeded;
+              });
         }
 
-        if (_max_size_exceeded) {
-            vlog(
-              stlog.warn,
-              "Log storage usage {} exceeds configured max {}",
-              human::bytes(usage.usage.total()),
-              human::bytes(_log_storage_max_size().value()));
+        if (_high_watermark_exceeded || _max_size_exceeded) {
```

Comment:
If we think of "max size" as "target size", then I don't think we need this whole "start trimming at 80%" logic at all.  We already reserve a percentage or fixed size of the disk globally, so we're kind of double-counting if we reserve 10% of disk globally, and then within raft effectively also reserve 20% because we start trimming proactively at 80% usage.

If we define the global thresholds to be something like "warn at 80%, block at 90%", then that 10% gap becomes the space within which we absorb whatever amount the raft storage overshoots its target by.  That "capacity slush fund" also accomodates any overshoots resulting from the walltime gap between the resource management code re-evaluating target capacities.